fletcher/_algorithms.py

```diff
@@ -65,6 +65,123 @@ def extract_isnull_bytemap(chunked_array):
     return result
 
 
+@numba.jit(nogil=True, nopython=True)
+def _argsort_insertionsort(indices, sa, lo, hi):
+    if hi <= lo:
+        return
+
+    for i in range(lo + 1, hi + 1):
+        e = indices[i]
+        j = i
+        if sa.elements_lt(e, indices[lo]):
+            while j > lo:
+                indices[j] = indices[j - 1]
+                j -= 1
+            indices[lo] = e
+        else:
+            while sa.elements_lt(e, indices[j - 1]):
+                indices[j] = indices[j - 1]
+                j -= 1
+            indices[j] = e
+
+
+@numba.jit(nogil=True, nopython=True)
+def _argqsort_string_array(indices, sa, lo, hi):
+    """
+    Quicksort implementation derived from the qsort algorithm outlined in
+
+    Mehlhorn, Kurt, and Peter Sanders. Algorithms and data structures:
+    The basic toolbox. Springer Science & Business Media, 2008.
+    """
+    while hi - lo + 1 > 40:
+        # Original states
+        # j = pickPivot
+        # swap(indices[j], indices[l])
+        # We pick p = lo for simplicity
+        p = indices[lo]
+        i = lo
+        j = hi
+        while True:
+            while sa.elements_lt(indices[i], p):
+                i += 1
+            while sa.elements_lt(p, indices[j]):
+                j -= 1
+            if i <= j:
+                tmp = indices[i]
+                indices[i] = indices[j]
+                indices[j] = tmp
+                i += 1
+                j -= 1
+            if i > j:
+                break
+        if i < ((lo + hi) / 2):
+            _argqsort_string_array(indices, sa, lo, j)
+            lo = i
+        else:
+            _argqsort_string_array(indices, sa, i, hi)
+            hi = j
+    _argsort_insertionsort(indices, sa, lo, hi)
```

Comment:
Without having the book at hand, I guess if `hi - lo + 1 > 40` doesn't hold, it is faster to use insertionsort? I think this might be worth mentioning since the book won't always be available. Also, it's not that obvious when the algorithm actually terminates.